packages/lesswrong/server/resolvers/bookmarksFeedResolver.ts

```diff
@@ -0,0 +1,165 @@
+    feedPost: FeedPost     
+    feedCommentThread: FeedCommentThread 
+    sortKey: Date!         
+  }
+  
+  extend type Query {
+    BookmarksFeed(
+      limit: Int,
+      cutoff: Date,
+      offset: Int,
+    ): BookmarksFeedQueryResults!
+  }
+`
+
+interface FeedCommentsThreadResolverType extends UltraFeedCommentsThreadResolverType {
+  post: Partial<DbPost> | null;
+}
+
+type InternalFeedItem = {
+  type: 'feedPost' | 'feedCommentThread';
+  feedPost: FeedPostResolverType | null;
+  feedCommentThread: FeedCommentsThreadResolverType | null;
+  sortKey: Date;
+};
+
+interface BookmarksFeedQueryResult {
+  cutoff: Date | null;
+  endOffset: number;
+  results: BookmarksFeedEntryType[];
+}
+
+export const bookmarksFeedGraphQLQueries = {
+  BookmarksFeed: async (_root: void, args: any, context: ResolverContext): Promise<BookmarksFeedQueryResult> => {
+    const {limit = 20, cutoff, offset = 0} = args;
+    const {currentUser} = context;
+    
+    if (!currentUser) {
+      throw new Error("Must be logged in to view bookmarks feed");
+    }
+
+    const selector: MongoSelector<DbBookmark> = {
+      userId: currentUser._id,
+      cancelled: false,
+    };
+    if (cutoff) {
+      selector.lastUpdated = {$lt: cutoff}; 
+    }
+
+    const bookmarks = await Bookmarks.find(selector, {
+      sort: {lastUpdated: -1},
+      limit: limit + 1, 
+      ...(offset && !cutoff ? { skip: offset } : {}) 
+    }).fetch();
+
+    const hasMore = bookmarks.length > limit;
+    const resultsBookmarks = hasMore ? bookmarks.slice(0, limit) : bookmarks;
+    const nextCutoff = hasMore ? (resultsBookmarks[resultsBookmarks.length - 1]?.lastUpdated ?? new Date()) : null;
+
+    const postBookmarkIds: string[] = [];
+    const commentBookmarkIds: string[] = [];
+    const bookmarkMap = new Map<string, DbBookmark>();
+
+    resultsBookmarks.forEach(bookmark => {
+      if (bookmark.documentId) {
+        bookmarkMap.set(bookmark.documentId, bookmark);
+        if (bookmark.collectionName === 'Posts') {
+          postBookmarkIds.push(bookmark.documentId);
+        } else if (bookmark.collectionName === 'Comments') {
+          commentBookmarkIds.push(bookmark.documentId);
+        }
+      }
+    });
+
+    const [posts, comments] = await Promise.all([
+      postBookmarkIds.length ? loadByIds(context, "Posts", postBookmarkIds) : Promise.resolve([]),
+      commentBookmarkIds.length ? loadByIds(context, "Comments", commentBookmarkIds) : Promise.resolve([])
+    ]);
+    
+    const commentPostIds = filterNonnull(comments).map(c => c.postId);
+    const validCommentPostIds = filterNonnull(commentPostIds);
+    const commentPosts = validCommentPostIds.length ? await loadByIds(context, "Posts", uniq(validCommentPostIds)) : [];
+    const commentPostsById = new Map(filterNonnull(commentPosts).map(p => [p._id, p]));
+
+    const feedItems: InternalFeedItem[] = [];
+
+    filterNonnull(posts).forEach(post => {
```

Comment:
these two `forEach`s should probably both just be .map(...).filter(...), but whatever